tests/simple/testdata/optionals.textproto

```diff
@@ -0,0 +1,237 @@
+name: "optionals"
+description: "Tests for optionals."
+section: {
+  name: "optionals"
+  test {
+    name: "none_or_none_or_value"
+    expr: "optional.none().or(optional.none()).orValue(42)"
+    value: { int64_value: 42 }
+  }
+  test {
+    name: "none_optMap_hasValue"
+    expr: "optional.none().optMap(y, y + 1).hasValue()"
+    value: { bool_value: false }
+  }
+  test {
+    name: "empty_map_optFlatMap_hasValue"
+    expr: "{}.?key.optFlatMap(k, k.?subkey).hasValue()"
+    value: { bool_value: false }
+  }
+  test {
+    name: "map_empty_submap_optFlatMap_hasValue"
+    expr: "{'key': {}}.?key.optFlatMap(k, k.?subkey).hasValue()"
+    value: { bool_value: false }
+  }
+  test {
+    name: "map_submap_subkey_optFlatMap_value"
+    expr: "{'key': {'subkey': 'subvalue'}}.?key.optFlatMap(k, k.?subkey).value()"
+    value: { string_value: "subvalue" }
+  }
+  test {
+    name: "map_submap_optFlatMap_value"
+    expr: "{'key': {'subkey': ''}}.?key.optFlatMap(k, k.?subkey).value()"
+    value: { string_value: "" }
+  }
+  test {
+    name: "map_optindex_optFlatMap_optional_ofNonZeroValue_hasValue"
+    expr: "{'key': {'subkey': ''}}.?key.optFlatMap(k, optional.ofNonZeroValue(k.subkey)).hasValue()"
+    value: { bool_value: false }
+  }
+  test {
+    name: "optional_of_optMap_value"
+    expr: "optional.of(42).optMap(y, y + 1).value()"
+    value: { int64_value: 43 }
+  }
+  test {
+    name: "optional_ofNonZeroValue_or_optional_value"
+    expr: "optional.ofNonZeroValue(42).or(optional.of(20)).value() == 42"
+    value: { bool_value: true }
+  }
+  test {
+    name: "ternary_optional_hasValue"
+    expr: "(has({}.x) ? optional.of({}.x) : optional.none()).hasValue()"
+    value: { bool_value: false }
+  }
+  test {
+    name: "map_optindex_hasValue"
+    expr: "{}.?x.hasValue()"
+    value: { bool_value: false }
+  }
+  test {
+    name: "has_map_optindex"
+    expr: "has({}.?x.y)"
+    value: { bool_value: false }
+  }
+  test {
+    name: "has_map_optindex_field"
+    expr: "has({'x': {'y': 'z'}}.?x.y)"
+    value: { bool_value: true }
+  }
+  test {
+    name: "type"
+    expr: "type(optional.none()) == optional_type"
+    value: { bool_value: true }
+  }
+  test {
+    name: "spaghetti_1"
```

Comment:
Even though they are spaghetti, let's call this `optional_chaining_#`

Reply:
Done.